Add Automata.makeCharSet/makeCharClass to optimize regexp

rmuir opened the pull request:
Previously caseless matching was implemented via code such as this:

```java
  Operations.union(Automata.makeChar('x'), Automata.makeChar('X'))
```

Proposed unicode caseless matching (#14192) implements it with repeated unions:

```java
  a1 = Operations.union(Automata.makeChar('x'), Automata.makeChar('X'))
  a2 = Operations.union(a1, Automata.makeChar('y'))
  a3 = Operations.union(a2, Automata.makeChar('Y'))
```
The union operation doesn't return a minimal automaton: improving union would always be nice, but this change offers a simple api for the task that returns half the number of states.

Before: caseless match of "a":
![a_before](https://github.com/user-attachments/assets/5f1f7cc2-a792-434f-98ec-a5e629678f16)

After:
![a_after](https://github.com/user-attachments/assets/aa60f883-cfd4-4291-9fb7-1bbecf0825ea)

Before: caseless match of "lucene":
![lucene_before](https://github.com/user-attachments/assets/76e38dc6-5cb4-4b77-923a-22dbab345241)

After:
![lucene_after](https://github.com/user-attachments/assets/3ca7755e-826d-4f79-b22e-8a514497d69e)

Just like the `union`, the `concatenate` adds some useless states, but they are less of a problem than the ones from before.

I didn't try anything more such as repeated union or kleene star, to see if I could make a really bad case, I felt like this was good enough, to get it to a better place. We can look at optimizing union/concatenate separately still, but that's always more dangerous and tricky.

rmuir committed:
Add Automata.makeCharSet(int[]) to optimize caseless matching.

Previously caseless matching was implemented via code such as this:

  Operations.union(Automata.makeChar('x'), Automata.makeChar('X'))

Proposed unicode caseless matching implements it with repeated unions:

  a1 = Operations.union(Automata.makeChar('x'), Automata.makeChar('X'))
  a2 = Operations.union(a1, Automata.makeChar('y'))
  a3 = Operations.union(a2, Automata.makeChar('Y'))

The union operation doesn't return a minimal automaton: improving union
would always be nice, but this change offers a simple api for the task
that returns half the number of states.

rmuir commented:
also `makeCharUnion()` comes to mind as a compelling alternative name, since there is already a `makeStringUnion()`. Naming is hard. just want to get the idea out there, since caseless regex is a common usecase.

jpountz reviewed (approved):
This looks good to me. Worst-case scenario, we make this new API delegate to union() when we improve it. I'm curious if you considered passing the code points as var-args for convenience?

rmuir commented:
I considered it but then didn't use any varargs after Dawid's email about compiler performance problems coming from them.

rmuir committed:
generalize makeCharSet(int[]) to makeCharClass(int[],int[])
* Add new "character class" node, which was previously composed by union
  of many nodes.
* Remove "predefined class" node, which previously built an internal
  separate regex on the fly, it is just another character class.
* RegExp no longer uses union() internally, except for union (|) operator.

rmuir commented:
I generalized this to `makeCharClass(int[],int[])`, added a "character class" node to use it instead of unioning many nodes, replaced the pre-built class functionality with it too.

The build failure is bogus: tests pass. It seems to be bug in compiler:
```
An exception has occurred in the compiler (21.0.5). Please file a bug against the Java compiler via the Java bug reporting page (https://bugreport.java.com/) after checking the Bug Database (https://bugs.java.com/) for duplicates. Include your program, the following diagnostic, and the parameters passed to the Java compiler in your report. Thank you.
```

rmuir commented:
That's error-prone that's broke trying to do some null analysis :)

rmuir commented:
anyway, I think this is the right path, rather than fight with union(), let's just get it out of our way. with this change union() is only used for union operator (`|`) and not internally.

rmuir commented:
My example for this one, if you have something like `[^a-gklM-O\s]`, with the case-insensitive flag maybe, it just calls the new `makeCharClass(int[],int[])` method and you get minimal automaton with the ranges you expect from state1->state2. No more walking up and down the parse tree and collecting sub-automata into lists and unioning them, no more sub-regexp creation and unioning for prebuilt classes.

rmuir commented:
I will figure out what angers the error-prone tomorrow. I am rusty with java, so this PR needs assistance LOL. but all the tests pass.

rmuir commented:
a few more notes:
* maybe we should deprecate `union(Automaton, Automaton)` and only leave `union(List<Automaton>)`. I see the former approach has proven trappy, let's guide developers to do it the faster way? The deprecation warning itself can be useful here to explain why. Nothing outside of tests uses this signature anymore.
* by having a way to create a character class node, if we wanted, we could give it additional boolean `@param complement True if the character class should be negated`. It would not help performance but would allow us to cleanup on the API side: the problematic complement() is needed and awesome for negated character classes, but the general algorithm is not sustainable and requires determinization. It is a crazy trap: complement() on one of these simple character classes is the perfect algorithm: totalize() to one dead state, then nuke the previous accept state. I definitely wrote the code and deleted it several times to just have this flag, to remove the general-purpose complement() trap elsewhere, which is not efficient at all for all situations. 
* the PR needs more tests but I fought the existing toString-based ones to exhaustion already. At least tests work and confirm it does what we expect
* warning: there may be dead code, that we could remove that I missed. We should look and try to fix RegExp class to not do everything package-private, because it makes it impossible for our tooling to detect dead code. If it used private correctly, tools would tell us if a function is unused by anything.

dweiss reviewed (approved):
This looks great. The only thing that caught my attention would be parallel arguments (from/to) - we could replace it with the more high-level and perhaps more obvious List<record CodepointRange>. But it's such a low-level class that perhaps we shouldn't try to be that fancy and just keep it as is... It's clear what the code is doing (to me at least).

rmuir commented:
I tried to update the error prone to fix its bugs, it is angry about the way we do gradle. I will YOLO my way thru this stuff. 

> The default --should-stop=ifError policy (INIT) is not supported by Error Prone, pass --should-stop=ifError=FLOW instead

rmuir commented:
Like i literally have no idea what this tool is trying to tell me there. But I think errorprone is broken, it depends on too many internals of the java compiler apis.

rmuir committed:
Merge branch 'main' into optimize_caseless_regexp

rmuir commented:
I would propose replacing this checker with ast-grep rules for whatever we need: it is not a good one. the use of internal java APIs is too crazy.

rmuir commented:
I opened https://github.com/apache/lucene/issues/14200 for the error-prone situation

rmuir committed:
Revert "disable error-prone as it is failing CI builds with NoSuchMethodError"

This reverts commit 4cd4f8e2f99b8e300fdd996b54920625a257acf7.

mikemccand commented on lucene/core/src/java/org/apache/lucene/util/automaton/RegExp.java:
```diff
@@ -648,13 +645,16 @@ private Automaton toAutomaton(
         break;
       case REGEXP_CHAR_RANGE:
-        a = Automata.makeCharRange(from, to);
+        a = Automata.makeCharRange(from[0], to[0]);
```
Do we check that `from.length == 1` and `to.length == 1` somewhere?

rmuir replied:
I looked. toString is a problem. so I want to leave this one be for now. To me it also doesn't make sense to keep adding more "variables" to the RegExp class , it has too many already for these nodes.

mikemccand commented on lucene/core/src/java/org/apache/lucene/util/automaton/Automata.java:
```diff
@@ -140,6 +141,32 @@ public static Automaton makeCharRange(int min, int max) {
   }
 
+  /** Returns a new minimal automaton that accepts any of the provided codepoints */
+  public static Automaton makeCharSet(int[] codepoints) {
```
I guess it's harmless if you have the same codepoint more than once in the incoming array?  Automaton will dedup those dups in `a.finishState()`?  Maybe we should add a test case making sure that's OK ...

rmuir replied:
I have a test for dups :) I don't have many tests but I tested that one.

mikemccand commented on lucene/core/src/java/org/apache/lucene/util/automaton/RegExp.java:
```diff
@@ -828,13 +838,10 @@ void toStringBuilder(StringBuilder b) {
   }
 
-  /** Like to string, but more verbose (shows the higherchy more clearly). */
+  /** Like to string, but more verbose (shows the hierarchy more clearly). */
```
Oooh nice typo fix!

mikemccand commented on lucene/core/src/java/org/apache/lucene/util/automaton/RegExp.java:
```diff
@@ -1050,14 +1059,25 @@ static RegExp makeDeprecatedComplement(int flags, RegExp exp) {
           "invalid range: from (" + from + ") cannot be > to (" + to + ")");
-    return newLeafNode(flags, Kind.REGEXP_CHAR_RANGE, null, 0, 0, 0, 0, from, to);
+    return newLeafNode(
+        flags, Kind.REGEXP_CHAR_RANGE, null, 0, 0, 0, 0, new int[] {from}, new int[] {to});
```
Oh if this is the only place where we create `REXEGP_CHAR_RANGE` (the API is not public) then we don't need to otherwise check `from.length == 1` and `to.length == 1`.

rmuir replied:
yeah it is. The only reason I didn't nuke these RANGE and CHAR nodes is ... toString()!

rmuir committed:
Merge branch 'main' into optimize_caseless_regexp

mikemccand commented on lucene/core/src/java/org/apache/lucene/util/automaton/RegExp.java:
```diff
@@ -1195,60 +1215,132 @@ final RegExp parseCharClassExp() throws IllegalArgumentException {
+      ends.add('A' - 1);
+      starts.add('Z' + 1);
+      ends.add('_' - 1);
+      starts.add('_' + 1);
+      ends.add('a' - 1);
```
Heh this range is just one character: ` (back quote).

mikemccand commented on lucene/core/src/java/org/apache/lucene/util/automaton/RegExp.java:
```diff
@@ -1195,60 +1215,132 @@ final RegExp parseCharClassExp() throws IllegalArgumentException {
+      ends.add((int) ' ');
+      next();
+    } else if (peek("S")) {
+      // non-whitespace: [^ \t\n\r]
```
Hmm maybe fix this comment to put the space char at the end for consistency (the chars are sorted by their ascii/codepoint, ascending)?

mikemccand commented on lucene/core/src/java/org/apache/lucene/util/automaton/RegExp.java:
```diff
@@ -1195,60 +1215,132 @@ final RegExp parseCharClassExp() throws IllegalArgumentException {
+      // non-whitespace: [^ \t\n\r]
+      starts.add(Character.MIN_CODE_POINT);
+      ends.add('\t' - 1);
+      starts.add('\n' + 1);
```
Maybe add a comment that you do not do the `\t-\n` case because they are adjacent codepoints (`0x9` and `0xa`) so there are no characters in the negated range in between them?

mikemccand reviewed (approved):
This looks great!  I left minor comments... thanks @rmuir.

Crazy how union makes dead states like that, but leave that battle for another day...

rmuir committed:
fix error-prone issue!

rmuir committed:
fix comment, so that it is obvious that it is a range. fix ordering to match both cases.

rmuir commented:
OK I tried out a List-based API  as alternative to array-based API. It isn't fully correct, which is part of my issue, but see it here: https://github.com/apache/lucene/commit/8b535a1c2fb4008dbe4e0b007ac95155dfe640ba

I was excited that it might cleanup the code, but it presented some problems. I think most problems are because we are talking about `int` arrays (codepoint ranges), but common use cases will of course use java `char`. 

array-based API:
```java
Automata.makeCharSet(new int[] {'a', 'A'});
```

List-based API:
```java
Automata.makeCharSet(List.of((int) 'a', (int) 'A'));
```

It requires user to make casts because the boxed types in java suck.

Also I'm not happy that with the List-based API, error-prone keeps catching me doing this:
```
error: [BoxedPrimitiveEquality] Comparison using reference equality instead of value equality. Reference equality of boxed primitive types is usually not useful, as they are value objects, and it is bug-prone, as instances are cached for some values but not others.
          if (from.get(i) == to.get(i)) {
                          ^
    (see https://errorprone.info/bugpattern/BoxedPrimitiveEquality)
  Did you mean 'if (Objects.equals(from.get(i), to.get(i))) {' or 'if (from.get(i).equals(to.get(i))) {'?
```

It happens because I keep forgetting to call `.intValue()` in special places.

So, I'd like to stick with array API, net-net I think it is better. I will look at messy List code that I have in this parser and see if I can improve that to reduce chances of problems when parsing lists of ranges. I also have other changes in that commit that are unrelated good ones and will poach them into here.

rmuir committed:
format codepoints in the internal parse tree output with U+%04X

Does not change the behavior of toString(), which should really be
revisited. This is just the internal parse tree used by tests, let's
make this one easy at least.

rmuir committed:
add sanity checks and ROOT

rmuir committed:
add test for a more complicated char class, that uses various features

rmuir committed:
test negated char class, make sure it comes back minimal

rmuir committed:
add case-insensitive tests.

Only the individual characters, not ranges, in a [] class are treated as
case-insensitive: this is the previous behavior unchanged. test it.

rmuir committed:
Fix concatenate to remove the dead states it creates, just like
intersection/union/etc do

This one doesn't take care, it adds dead states to basically every
regexp. With the change, many regexp are now minimal.

It runs in linear time so it doesn't change the complexity of
concatenate() but just cleans up the mess.

rmuir commented:
Finally! The concatenate() issue was an easy fix, it neglected to clean up its dead states. All of its partners in crime do this, but the fact we neglect it for concatenate messes up too many As.

![Screen_Shot_2025-02-05_at_23 55 55](https://github.com/user-attachments/assets/f5b5234c-73cd-4431-abed-6f39023eb687)

rmuir commented:
I'm feeling good about this one now, with the change, a lot of regexps now come out minimal from the start, which is a good thing. 

We also eliminate overhead of tons of nodes, which is important if we ever want to support caseless range matches (e.g. `[a-z]` with case-insensitive flag matching 'A'). To implement it generally (not just for ascii), we need to iterate the range and add tons of alternatives, but its an `int[]` and will all be folded into a single state: basically as good as we can get.

rmuir committed:
fix dead-states-test to explicitly create dead states, rather than
relying on some function to create a mess. it doesn't anymore.

rmuir commented on lucene/core/src/test/org/apache/lucene/util/automaton/TestAutomaton.java:
```diff
@@ -667,11 +667,14 @@ public void testConcatenatePreservesDet() throws Exception {
     a = Operations.removeDeadStates(a);
-    assertEquals(3, a.getNumStates());
+    assertEquals(1, a.getNumStates());
   }
```
cc @jpountz this is how i fixed this test.

jpountz replied:
Much better, thanks.

dweiss commented:
> Finally! The concatenate() issue was an easy fix, it neglected to clean up its dead states. All of its partners in crime do this, but the fact we neglect it for concatenate messes up too many As.
> 
> ![Screen_Shot_2025-02-05_at_23 55 55](https://private-user-images.githubusercontent.com/504194/410297736-f5b5234c-73cd-4431-abed-6f39023eb687.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzg4NTk0NjQsIm5iZiI6MTczODg1OTE2NCwicGF0aCI6Ii81MDQxOTQvNDEwMjk3NzM2LWY1YjUyMzRjLTczY2QtNDQzMS1hYmVkLTZmMzkwMjNlYjY4Ny5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMjA2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDIwNlQxNjI2MDRaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1lNWZhYjljM2Q2YzlkMjBjNWYyMDMzNDkxMjQwN2ZjMDE1MGFhYzgxYmFmODMwM2VjNTE4ZWMwZWE1ODJlMDlhJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.Rh69pJtLL-XnnnprIgmM_nlBl6KxABbd2lRpcg0kdAU)

God this is so much nicer compared to that first/ original automaton you included!
